src/pytorch_lightning/utilities/hpu_datamodule.py

```diff
@@ -0,0 +1,170 @@
+# Copyright The PyTorch Lightning team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Any, Optional
+
+import pytorch_lightning as pl
+from pytorch_lightning.utilities import _HPU_AVAILABLE
+from pytorch_lightning.utilities.imports import _TORCHVISION_AVAILABLE
+
+if _TORCHVISION_AVAILABLE:
+    import torchvision.datasets
+    from torchvision import transforms as transform_lib
+
+if _HPU_AVAILABLE:
+    try:
+        import habana_dataloader
+    except ImportError:
+        raise ModuleNotFoundError("habana_dataloader package is not installed.")
+
+_DATASETS_PATH = "./data"
+
+
+def load_data(traindir, valdir, train_transforms, val_transforms):  # type: ignore[no-untyped-def]
+
+    # check supported transforms
+    if train_transforms is not None:
+        for t in train_transforms:
+            if (
+                isinstance(t, transform_lib.RandomResizedCrop)
+                or isinstance(t, transform_lib.RandomHorizontalFlip)
+                or isinstance(t, transform_lib.ToTensor)
+                or isinstance(t, transform_lib.Normalize)
+            ):
+
+                continue
+            else:
+                raise ValueError("Unsupported train transform: " + str(type(t)))
+
+        train_transforms = transform_lib.Compose(train_transforms)
+
+    if val_transforms is not None:
+        for t in val_transforms:
+            if (
+                isinstance(t, transform_lib.Resize)
+                or isinstance(t, transform_lib.CenterCrop)
+                or isinstance(t, transform_lib.ToTensor)
+                or isinstance(t, transform_lib.Normalize)
+            ):
+
+                continue
+            else:
+                raise ValueError("Unsupported val transform: " + str(type(t)))
+
+        val_transforms = transform_lib.Compose(val_transforms)
+
+    if "imagenet" not in traindir.lower() and "ilsvrc2012" not in traindir.lower():
+        raise ValueError("Habana dataloader only supports Imagenet dataset")
+
+    # Data loading code
+    dataset_train = torchvision.datasets.ImageFolder(traindir, train_transforms)
+
+    dataset_val = torchvision.datasets.ImageFolder(valdir, val_transforms)
+
+    return dataset_train, dataset_val
+
+
+class HPUDataModule(pl.LightningDataModule):
```

Comment:
IMO we shouldn't add this here. There are multiple reasons:

1.) This is too domain-specific. It only allows CV datasets (more specifically only imagenet so far). Lightning tries to be agnostic w.r.t. the task and data domain at hand.

2.) It is too accelerator specific. We want all the code to be hardware specific for the user and switching out the DM depending on the hardware isn't really agnostic. Usually hardware only reflects to accelerators and if necessary strategies and plugins. Personally, I'd prefer to keep this limitation.

3.) There are a ton of tools to accelerate loading for a specific usecase. There is nvidia dali, ffcv and many others. The interface of PL allows you to easily plugin each of those without requiring any additional code on our side. As far as I can tell, this would be the same case with the `habana_dataloader`. 

4.) IMO this does not really add a lot of benefits to the average user (mainly due to (1)) but only moves the habana imports to lightning imports.

Reply:
Thanks @justusschock for the quick review. Points to consider
- there needs to be a place to maintain the extensible classes of datamodule per accelerator. Lightning should provide interfaces to support generic as well as accelerator specific parameters.
- I view datamodule as one more strategy/ plugins wrt accelerators. It need not be generic. If all strategies needs to be generic, we need to have a mechanism to ensure all plugins/ strategies work well with all available accelerators.
- AFAIK, dali doesn't support all datasets. We intend to enhance the supported data formats in the subsequent releases
- We need a mechanism in lightning to show case users on how to make use specific features. Error handling as part of it is only to make user life simple

Reply:
I agree with @justusschock. We don't want to add a `HPUDataModule` just as there isn't an `HPULightningModule` or `HPUTrainer`. If you want the implementation to live in this repository, we can choose one of the following two options instead:
- The strategy manages the DataLoader, as we do for IPUs (hacky solution but simple): https://github.com/Lightning-AI/lightning/blob/2b7cd58e06702e7fc96e65001a3853ed2aabc1bf/src/pytorch_lightning/strategies/ipu.py#L240-L254
- Implement a new mechanism to add plug-in support for custom DataLoaders, this would be the same for HPUs as for FFCV etc. For example, a mixin class. This will require a design discussion.

If you still prefer the simplicity of the HPUDataModule, this is totally fine but we would want that it is maintained inside the `hpu_dataloader` library. Where we update our docs to suggest it to HPU users.

Reply:
Thanks @carmocca . 
- I think a generic design is needed so that all accelerators can inherit & it create a new strategy for datamodule. I believe the proposed PR partially resolves it. 
- Poptorch method doesn't seem scalable.
- The changes cannot be maintained in hpu_dataloader as its specific to lightning plugins. 
- We already showcase pytorch users on how to make use of habana_Dataloader. I prefer a similar mechanism for lightning users. Just updating the document will not help a simple user I think.